### PR TITLE
[#185, #186] Create embedder class, tensors caching

### LINKED ISSUE #185: Create embedder class
Create Embedder class that encapsulates all embedders-related logic (loading, using, tensors precalculation) in a convenient way.

Some related functions:

- use_service_embedder
- use_user_embedder

Function proposition (food for thought):

- load_last_embedder

### LINKED ISSUE #186: Tensors caching
- [ ] Specify models (in the DB) that should have cached tensors.
- [ ] Specify fields name.
- [ ] Provide implementation of tensor caching throughout the whole project.

### PR DESCRIPTION
**Pull request by JanKapala**
Close #185, close #186

### CODE REVIEW

File: recommender/engine/training/autoencoders.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ def evaluate_autoencoder(model, dataloader, loss_function, device):
     model.eval()
     with torch.no_grad():
         for batch in dataloader:
-            features = batch[0].to(device)
+            features = batch[0].float().to(device)
```
Comment (wujuu) [inline recommender/engine/training/autoencoders.py:19]: Shouldn't we be casting to float earlier, in the dataloader for example?

File: recommender/errors.py [hunk 2]
```diff
@@ -46,3 +46,23 @@ def message(self):
...
+    def message(self):
+        return "One or more objects don't have one hot tensor"
+
+
+class MissingDenseTensorError(RecommendationEngineError):
+    def message(self):
+        return "One or more objects don't have dense tensor needed to"
+
+
+class MissingDenseTensorError(RecommendationEngineError):
```
Comment (wujuu) [inline recommender/errors.py:66]: Repeated the same error class as implemented above

File: tests/engine/preprocessing/test_embeder.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+from recommender.errors import MissingDenseTensorError, MissingOneHotTensorError
+from tests.factories.marketplace import ServiceFactory
+
+
+@pytest.fixture
+def services(mongo):
+    return [ServiceFactory(id=i) for i in range(10)]
+
+
+def test_embedder(mongo, services):
```
Comment (wujuu) [inline tests/engine/preprocessing/test_embeder.py:17]: I know this might be inconvenient, but this test should be split into smaller test cases, for clarity. I would avoid writing very long tests that test everything and prefer smaller, well defined test cases - to refactor later

**Review by wujuu — CHANGES_REQUESTED** [review]

File: recommender/engine/agents/rl_agent/utils.py [hunk 4]
```diff
@@ -93,7 +93,7 @@ def create_itemspace(embedder: torch.nn.Module) -> Tuple[torch.Tensor, pd.DataFr
     return service_embedded_tensors, index_id_map
 
 
-def embedded_tensors_exist(objects: Union[Iterable, QuerySet]):
+def dense_tensors_exist(objects: Union[Iterable, QuerySet]):
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engine/agents/rl_agent/utils.py:96]: Maybe, but I don't exclude the possibility of using it without embedders.

File: recommender/engine/training/autoencoders.py (hunk 1 above)
Comment (JanKapala) [reply recommender/engine/training/autoencoders.py:19]: this is test hotfix and it will be refactored in other issue due to issues prioritization

File: tests/engine/preprocessing/test_embeder.py (hunk 3 above)
Comment (JanKapala) [reply tests/engine/preprocessing/test_embeder.py:17]: Godd suggestion but in this case each line is advancing previous setup, so next assertion uses changed test setup of the previous assertions. This is the only reason to have all of these assertion in one test case. If we are spliting this into multiple methods we would have to make a lot of repeated setup.

File: recommender/engine/agents/rl_agent/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (wujuu) [reply recommender/engine/agents/rl_agent/utils.py:96]: Yeah but they should be moved to the embedder module, maybe as a statis method even.

File: recommender/engine/agents/rl_agent/utils.py [hunk 5]
```diff
@@ -103,10 +103,25 @@ def embedded_tensors_exist(objects: Union[Iterable, QuerySet]):
...
     objects = list(objects)
-    first_len = len(objects[0].embedded_tensor)
+    first_len = len(objects[0].dense_tensor)
     if first_len == 0:
         return False
-    return all(len(obj.embedded_tensor) == first_len for obj in objects)
+    return all(len(obj.dense_tensor) == first_len for obj in objects)
+
+
+def one_hot_tensors_exist(objects: Union[Iterable, QuerySet]):
```
[the remark this answers is not in the document]
Comment (wujuu) [reply recommender/engine/agents/rl_agent/utils.py:112]: Yeah but they should be moved to the embedder module, maybe as a statis method even.

File: tests/engine/preprocessing/test_embeder.py (hunk 3 above)
Comment (JanKapala) [reply tests/engine/preprocessing/test_embeder.py:17]: But in this particular case it's good approach, because we have to use embedders to make scenario for other embedders tests (it is inevitable). We know exactly which functionality is failing because each assertion set is executed sequentially, so when one of them is failing you know that you have to first fix this assertion and not to fix next potentially failing assertions because after you fix the first one, next could be resolved too. If it will be not it doesn't matter, you should still fix the new one first failing assertion to make sure that it doesn't interfere with the next.

To sum it up: you will always know which part of test code has something to do with fixing when something is not working: the part from the beginning of the test to the failing assertion

Comment (wujuu) [reply tests/engine/preprocessing/test_embeder.py:17]: > But in this particular case it's good approach, because we have to use embedders to make scenario for other embedders tests (it is inevitable). We know exactly which functionality is failing because each assertion set is executed sequentially, so when one of them is failing you know that you have to first fix this assertion and not to fix next potentially failing assertions because after you fix the first one, next could be resolved too. If it will be not it doesn't matter, you should still fix the new one first failing assertion to make sure that it doesn't interfere with the next.
> 
> To sum it up: you will always know which part of test code has something to do with fixing when something is not working: the part from the beginning of the test to the failing assertion

This is true, and can stay that way, but you will agree that if you see that a particular, more granular test case is failing it is more clear what to fix and what is failing, than to check individual assertions) - this is also useful for checking for regression in the upcoming versions of the code

Comment (JanKapala) [reply tests/engine/preprocessing/test_embeder.py:17]: > > But in this particular case it's good approach, because we have to use embedders to make scenario for other embedders tests (it is inevitable). We know exactly which functionality is failing because each assertion set is executed sequentially, so when one of them is failing you know that you have to first fix this assertion and not to fix next potentially failing assertions because after you fix the first one, next could be resolved too. If it will be not it doesn't matter, you should still fix the new one first failing assertion to make sure that it doesn't interfere with the next.
> > To sum it up: you will always know which part of test code has something to do with fixing when something is not working: the part from the beginning of the test to the failing assertion
> 
> This is true, and can stay that way, but you will agree that if you see that a particular, more granular test case is failing it is more clear what to fix and what is failing, than to check individual assertions) - this is also useful for checking for regression in the upcoming versions of the code

sure, I think generaly spliting it in more tests is a good idea, but still I don't know how to do it gracefully in pytest - but definitely it can be considered in the next issue.

Comment (wujuu) [reply tests/engine/preprocessing/test_embeder.py:17]: > Yeah, in ruby, but how the gradual setup should be implemented here in pytest?

Using fixtures, you can do literally anything (including running some code, not only returning some objects and then tearing down the setup if needed, using the 'yield' keyword)

Comment (Michal-Kolomanski) [reply tests/engine/preprocessing/test_embeder.py:17]: I agree with @wujuu that the idea behind this test is a little bit hard to follow. Refactoring or solid description/documentation can be a solution. Of course, taking under consideration tasks priority.

**Review by Michal-Kolomanski — COMMENTED** [review]
Good job!
